cocotb/__init__.py

```diff
@@ -106,8 +107,38 @@ def _reopen_stream_with_buffering(stream_name):
 plusargs = {}
 """A dictionary of "plusargs" handed to the simulation."""
 
-# To save typing provide an alias to scheduler.add
-fork = scheduler.add
+
+def create_task(coro: Union[Coroutine, cocotb.decorators.RunningTask]) -> cocotb.decorators.RunningTask:
+    """
+    Runs task concurrently. Can be awaited at any point to block until a result is available.
+    """
+    return scheduler.add(coro)
+
+
+def fork(coro: Union[Coroutine, cocotb.decorators.RunningTask]) -> cocotb.decorators.RunningTask:
+    """
+    Runs task concurrently, but will end test if an exception occurs *before* being awaited.
+
+    Useful for "fire and forget" coroutines common in many testbench objects.
+    """
+    co = create_task(coro)
+
+    async def watcher():
+        try:
+            await co
+        except Exception as e:
+            if len(scheduler._scheduling) == 1:  # if we are the only thing waiting on the join
+                scheduler.finish_test(e)
+            else:
+                cocotb.utils.remove_traceback_frames(e, ['watcher', '__await__'])
```

Comment:
It's pretty annoying that you have to do this. Can you construct an equivalent case in `asyncio` or `trio`?

I'm remembering now that `trio.outcome`s had an  odd feature that prevented them being "sent" to more than one place. Perhaps this was the motivation.

Reply:
These are supposed to be asymmetric coroutines, being able to have multiple joins on the same coroutine is not kosher.

I'll work around this another way.

Reply:
I need a way to obtain a reference to a `Join` on the watcher coroutine to check within the watcher if there is a join on it. I think that means I'll have to make subclass of `RunningTask` specifically for `fork`.

```python
def fork(coro: Union[Coroutine, cocotb.decorators.RunningTask]) -> cocotb.decorators.RunningTask:
    """
    Runs a coroutine concurrently. Will end the current test if the coroutine results in an exception *before* being awaited.

    Preserved for legacy.
    """
    co = create_task(coro)

    async def watcher():
        try:
            return await co
        except Exception as e:
            cocotb.utils.remove_traceback_frames(e, ['watcher', '__await__'])
            if cocotb.triggers.Join(self) in scheduler._trigger2coros:
                warnings.warn(
                    "If you want to run a coroutine concurrently and want it to kill the test if there is an error, "
                    "prefer cocotb.fire_and_forget().")
                scheduler.finish_test(e)
            raise

    return create_task(watcher())
```

Reply:
We have no way to react to a `kill` here either... If the watcher is `kill`ed the coroutine it watches should be as well, so that's even more of a case for a a `RunningTask` subclass. We need to get #1349 in... That's a pretty critical feature IMO.

Reply:
This has been addressed.